### DISCUSSION TITLE
fs.copyFileSync and fs/promises.copyFile hangs

### DISCUSSION THREAD

**Discussion by Zain-ul-din** [nodejs/help#4975] [Q&A]
**Node Js Version : v16.16.0** 
**Plateform : Window**

Hey, I'm using node js to copy and paste files from one folder to another but, whenever I tried to copy a large number of files it hangs can you help me to solve this issue?

```typescript 
#!/usr/bin/env node
// Randoms CLI Working with files

import { existsSync, mkdirSync, copyFile } from "fs";
import { readdir, lstat,  } from "fs/promises";
import { extname } from "path";

const environment:string = "testing";

var root: string = process.cwd(); // absolute path
const folderPrefix: string = "RANDOM";
async function isFolder(path: string): Promise<boolean> {
  return (await lstat(path)).isDirectory();
}

function getFileName(filePath: string): string {
  const paths: Array<string> = filePath.split("\\", filePath.length);
  return paths[paths.length - 1];
}

async function addFileToRoot(filePath: string) {
  if (environment == "testing") return;
  const extName: string = extname(filePath).replaceAll(".", "");
  const folderName: string = (folderPrefix + "_" + extName).replaceAll(
    " ",
    "_"
  );
  var absFolderPath: string = root + "\\" + folderName;
  if (!existsSync(absFolderPath)) {
    mkdirSync(absFolderPath);
  }

  absFolderPath += "\\" + getFileName(filePath);
  console.log (absFolderPath);
  copyFile(filePath, absFolderPath, (err)=>{
    console.log (err);
  });
}

async function readDirectory(root: string): Promise<void> {
  if (!(await isFolder(root))) {
    addFileToRoot(root);
    return;
  }
  const files: string[] = await readdir(root);
  for (let i: number = 0; i < files.length; i += 1)
    readDirectory(root + "\\" + files[i]);
}

console.log ("Copying Files...");
setTimeout (()=> {
  console.log ("Making Dirs...");
} , 1000);
await readDirectory(root);
```
Here is the code I'm using I am pretty sure about the path is correct and also I'm not getting any error in the console.

**Comment by nagayev** [comment]
You can use `fs-extra`:
```js
const fse = require('fs-extra');

const srcDir = `path/to/file`;
const destDir = `path/to/destination/directory`;
                              
// To copy a folder or file  
const opts = {}; //{ overwrite: true } if you want to replace existing folder or file with same name
fse.copySync(srcDir, destDir, opts, function (err) {
  if (err) {                 
    console.error(err);
  } else {
    console.log("success!");
  }
});
```
Another way is using native utils like unix command `cp`.
This is probably the fastest way to copy files or directory.

**Reply by Zain-ul-din** [reply]
Is it work with typescript or not?

**Reply by nagayev** [reply]
@Zain-ul-din Yes, you can use it with typescript
https://github.com/jprichardson/node-fs-extra/blob/master/README.md#typescript

**Reply by nagayev** [reply]
@Zain-ul-din If I helped you please mark my reply as an answer. I'm trying to get github's Galaxy Brain archiment 😀.

**Reply by Zain-ul-din** [reply]
The syntax you recommend is working but the issue is still there however you do your job very well can we collaborate in the future and build something?

**Reply by nagayev** [reply]
@Zain-ul-din I think it's a great idea.
I'm looking for person to make projects together.

**Reply by Zain-ul-din** [reply]
so I'm here what do you wanna make?

**Reply by Zain-ul-din** [reply]
Anonymous762#2283 here is my discord account I think you'll contact me

**Reply by Zain-ul-din** [reply]
I know 4 to 5 programming languages you can choose any of these. C++, Javascript, typescript, c#, and also SQL and some non-SQL tools like firebase
Make repo and send me link I'll fork it

**Reply by Zain-ul-din** [reply]
I know python but i don't like it to much

**Reply by nagayev** [reply]
@Zain-ul-din Fine

**Reply by nagayev** [reply]
@Zain-ul-din Why you have remove your approvement?

**Reply by Zain-ul-din** [reply]
because your answer is not working!

**Reply by nagayev** [reply]
@Zain-ul-din What about using native utilities like unix `cp`?

**Reply by Zain-ul-din** [reply]
no, I'm window user

**Reply by nagayev** [reply]
@Zain-ul-din you can use Windows alternative - `copy`.
For details see
https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/copy

**Reply by Zain-ul-din** [reply]
Native APIs are slow and also not multi-threaded
